src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -795,15 +795,14 @@
       {:kind :invalid-subject-name
        :msg  (i18n/tru "Subject contains a wildcard, which is not allowed: {0}" subject)}))
 
-  (when (or (str/ends-with? subject ".")
-            (str/ends-with? subject "-"))
+  (when (str/ends-with? subject "-")
     (log/info (i18n/tru "Rejecting subject \"{0}\" as it ends with an invalid character" subject))
     (sling/throw+
      {:kind :invalid-subject-name
       :msg  (i18n/tru "Subject hostname format is invalid")}))
 
   (let [segments (str/split subject pattern-match-dot)]
-    (when-not (re-matches pattern-starts-with-alphanumeric (first segments))
+    (when-not (re-matches pattern-starts-with-alphanumeric-or-underscore (first segments))
```

Comment:
Is this code path used to validate the subject for the puppet root CA or intermediate CA, e.g. 
```
        Subject: CN=Puppet CA: anaerobic-dram.delivery.puppetlabs.net
```

Reply:
Any validation of the CA subjects should be over in the `puppetserver ca setup` or `puppetserver ca import` actions:

https://github.com/puppetlabs/puppetserver-ca-cli/tree/main/lib/puppetserver/ca/action

And is done while Puppet Server is offline because CA needs to be created, and a Puppet client certificate issued, in order for Puppet Server to start and serve the CA API.

Reply:
This path is only for agent certs, and not CAs.